File: Tarea 3/diet.py
from sys import stdin

def solve(A, n, x, mem):
    ans = 0
    if((n, x) in mem):
        ans = mem[(n, x)]
    else:
        if(n == 0):
            ans = -float('inf')
        elif(x <= 0):
            ans = x
        else:
            ans = max(solve(A, n-1, x - (A[n-1]), mem), solve(A, n-1, x, mem))
        mem[(n, x)] = ans
    return ans

def main():
    C = int(stdin.readline())
    while(C != 0):
        mem = {}
        calories = int(stdin.readline())
        tam = int(stdin.readline())
        A = list(map(int, input().split()))
        if(sum(A) > calories):
            print(f"{calories + (solve(A, tam, calories, mem) * -1)}")
        else:
            print("NO SOLUTION")
        C-=1

main()

File: Tarea 3/diet.cpp
#include <iostream>
#include <map>
#include <vector>
#include <tuple>
#include <stdio.h>
#include <string.h>
#include <algorithm>
#include <limits>
#define ini vector<int>
int dp[105][2505];
#define INF -2147483648;
using namespace std;

int solve(ini A, int n, int x)
{
    int ans = 0;

    if (dp[n][x] != -1)
    {
        ans = dp[n][x];
    }
    else{
        if(n == 0){
            ans = INF;
        }else if(x <= 0){
            ans = x;
        }else{
            ans = max(solve(A, n-1, x - (A[n-1])), solve(A, n-1, x));
        }
        dp[n][x] = ans;
    }
    return ans;
}
int main()
{
    int C, calories, tam = 0;
    cin >> C;
    while (C != 0)
    {
        memset(dp, -1, sizeof(dp));
        cin >> calories;
        cin >> tam;
        ini A;
        for (int i = 0; i < tam; i++)
        {
            cin >> A[i];
        }

        printf("%d", calories + (solve(A, tam, calories) * -1));
        C--;
    }
}